File: ocyaml.c
/* OCaml bindings for libyaml                                          */
/* (c)Copyright Aesthetic Integration, Ltd., 2016                      */
/* All rights reserved.                                                */
/*                                                                     */
/* Released under Apache 2.0 license as described in the file LICENSE. */
/*                                                                     */
/* Contributors:                                                       */
/* Konstantin kanishchev (kostya@aestheticintegration.com)             */
/*                                                                     */


#include <caml/mlvalues.h>
#include <caml/alloc.h>
#include <caml/memory.h>
#include <caml/misc.h>
#include <caml/signals.h>

const int Scalar = 0; 
const int Collection = 1;
const int Structure = 2;

#include <stdio.h>
#include <yaml.h>


CAMLprim value read_yaml(yaml_parser_t * p_parser);

CAMLprim value 
read_structure(yaml_parser_t * p_parser)
{
    CAMLparam0();
    CAMLlocal4(res, hd, tl, pair);
    tl = Val_emptylist;
    yaml_token_t token;   
    do {
        yaml_parser_scan(p_parser, &token);
        switch(token.type){
            case YAML_KEY_TOKEN:   
//                printf("Key \n");
                yaml_token_delete(&token);
                pair = caml_alloc(2,0);
                Store_field( pair, 0, read_yaml(p_parser) );  
                break;
            case YAML_VALUE_TOKEN: 
//                printf("Value \n");
                yaml_token_delete(&token);
                Store_field( pair, 1, read_yaml(p_parser) );  
                hd = caml_alloc(2,0);
                Store_field( hd, 0, pair);  
                Store_field( hd, 1, tl );              
                tl = hd;
                break;
            case YAML_BLOCK_END_TOKEN: 
//                printf("Mapping end \n");
                break;
            default:
                // TODO: Exception here
//                printf("Got unexpected token of type %d while reading a structure\n", token.type); fflush(stdout);
                yaml_token_delete(&token);
        }
    } while(token.type != YAML_BLOCK_END_TOKEN);
    yaml_token_delete(&token);
    res = caml_alloc(1, Structure);
    Store_field(res, 0, tl);
    CAMLreturn (res);
}

CAMLprim value 
read_sequence(yaml_parser_t * p_parser)
{
    CAMLparam0();
    CAMLlocal3(res, hd, tl);
    tl = Val_emptylist;
    yaml_token_t token;   
    do {
        yaml_parser_scan(p_parser, &token);
        switch(token.type){
            case YAML_BLOCK_ENTRY_TOKEN:   
//                printf("Block entry\n");
                yaml_token_delete(&token);
                hd = caml_alloc(2,0);
                Store_field( hd, 0, read_yaml(p_parser));  
                Store_field( hd, 1, tl );              
                tl = hd;
                break;
            case YAML_BLOCK_END_TOKEN: 
//                printf("Sequence end\n");
                break;
            default:
                // TODO: Exception here
//                printf("Got unexpected token of type %d while reading a sequence\n", token.type); fflush(stdout);
                yaml_token_delete(&token);
        }
    } while(token.type != YAML_BLOCK_END_TOKEN);
    yaml_token_delete(&token);
    res = caml_alloc(1, Collection);
    Store_field(res, 0, tl);
    CAMLreturn (res);
}

CAMLprim value 
read_yaml(yaml_parser_t * p_parser)
{
    CAMLparam0();
    CAMLlocal1(res);
    yaml_token_t token;   
    yaml_parser_scan(p_parser, &token);
    switch(token.type){
        case YAML_BLOCK_MAPPING_START_TOKEN:  
//            printf("Mapping start\n");
            yaml_token_delete(&token);
            res = read_structure(p_parser); 
            break;
        case YAML_BLOCK_SEQUENCE_START_TOKEN: 
//            printf("Sequence start\n");
            yaml_token_delete(&token);
            res = read_sequence (p_parser); 
            break;
        case YAML_SCALAR_TOKEN:     
//            printf("Scalar \n");
            res = caml_alloc(1, Scalar);
            Store_field(res, 0, caml_copy_string(token.data.scalar.value));
            yaml_token_delete(&token);
            break;
        default:
            // TODO: Exception here
//            printf("Got unexpected yaml token of type %d\n", token.type); fflush(stdout);
            yaml_token_delete(&token);
    }
    CAMLreturn (res);
}

CAMLprim value test (value v1)
{
    CAMLparam1(v1);
    CAMLlocal1(res);

    FILE *fh = fopen( String_val(v1), "r");
    yaml_parser_t parser;
    yaml_token_t  token;   

    // TODO: OCAML exceptions here
    if(!yaml_parser_initialize(&parser))
        fputs("Failed to initialize parser!\n", stderr);
    if(fh == NULL)
        fputs("Failed to open file!\n", stderr);

    yaml_parser_set_input_file(&parser, fh);
    yaml_parser_scan(&parser, &token);

    do {
        switch(token.type)
        {
            case YAML_STREAM_START_TOKEN: 
//                printf("Stream start\n");
                yaml_token_delete(&token);
                res = read_yaml(&parser); 
                CAMLreturn (res);
                break;
            case YAML_STREAM_END_TOKEN:
//                printf("Stream end\n");
                break;
            default:
            //    printf("Got unexpected toplevel token of type %d\n", token.type);
                yaml_token_delete(&token);
        }
    } while(token.type != YAML_STREAM_END_TOKEN);
    yaml_token_delete(&token);
    CAMLreturn (res);
}

File: ocyaml.ml
(* OCaml bindings for libyaml                                          *)
(* (c)Copyright Aesthetic Integration, Ltd., 2016                      *)
(* All rights reserved.                                                *)
(*                                                                     *)
(* Released under Apache 2.0 license as described in the file LICENSE. *)
(*                                                                     *)
(* Contributors:                                                       *)
(* Konstantin kanishchev (kostya@aestheticintegration.com)             *)
(*                                                                     *)

type yaml =
    | Scalar of string
    | Collection of yaml list
    | Structure of (yaml * yaml) list

(** Many bad things could happen with the parser and emitter. *)
type yaml_error_type =
    | YAML_NO_ERROR       (** No error is produced. *)
    | YAML_MEMORY_ERROR   (** Cannot allocate or reallocate a block of memory. *)
    | YAML_READER_ERROR   (** Cannot read or decode the input stream. *)
    | YAML_SCANNER_ERROR  (** Cannot scan the input stream. *)
    | YAML_PARSER_ERROR   (** Cannot parse the input stream. *)
    | YAML_COMPOSER_ERROR (** Cannot compose a YAML document. *)
    | YAML_WRITER_ERROR   (** Cannot write to the output stream. *)
    | YAML_EMITTER_ERROR  (** Cannot emit a YAML stream. *)


(** The stream encoding. *)
type encoding =
    | YAML_ANY_ENCODING     (* Let the parser choose the encoding. *)
    | YAML_UTF8_ENCODING    (** The default UTF-8 encoding. *)
    | YAML_UTF16LE_ENCODING (** The UTF-16-LE encoding with BOM. *)
    | YAML_UTF16BE_ENCODING (** The UTF-16-BE encoding with BOM. *)

(** Scalar styles. *)
type scalar_style =
    | YAML_ANY_SCALAR_STYLE (** Let the emitter choose the style. *)
    | YAML_PLAIN_SCALAR_STYLE (** The plain scalar style. *)
    | YAML_SINGLE_QUOTED_SCALAR_STYLE (** The single-quoted scalar style. *)
    | YAML_DOUBLE_QUOTED_SCALAR_STYLE (** The double-quoted scalar style. *)
    | YAML_LITERAL_SCALAR_STYLE (** The literal scalar style. *)
    | YAML_FOLDED_SCALAR_STYLE  (** The folded scalar style. *)


(** Token types. *)
type token =
    | YAML_NO_TOKEN
    | YAML_STREAM_START_TOKEN of encoding
    | YAML_STREAM_END_TOKEN
    | YAML_VERSION_DIRECTIVE_TOKEN of int * int
    | YAML_TAG_DIRECTIVE_TOKEN  of string * string
    | YAML_DOCUMENT_START_TOKEN
    | YAML_DOCUMENT_END_TOKEN
    | YAML_BLOCK_SEQUENCE_START_TOKEN
    | YAML_BLOCK_MAPPING_START_TOKEN
    | YAML_BLOCK_END_TOKEN
    | YAML_FLOW_SEQUENCE_START_TOKEN
    | YAML_FLOW_SEQUENCE_END_TOKEN
    | YAML_FLOW_MAPPING_START_TOKEN
    | YAML_FLOW_MAPPING_END_TOKEN
    | YAML_BLOCK_ENTRY_TOKEN
    | YAML_FLOW_ENTRY_TOKEN
    | YAML_KEY_TOKEN
    | YAML_VALUE_TOKEN
    | YAML_ALIAS_TOKEN of string
    | YAML_ANCHOR_TOKEN of string
    | YAML_TAG_TOKEN of string * string
    | YAML_SCALAR_TOKEN of scalar_style * string

let string_of_token = function
  | YAML_NO_TOKEN -> "YAML_NO_TOKEN"
  | YAML_STREAM_START_TOKEN _ -> "YAML_STREAM_START_TOKEN _"
  | YAML_STREAM_END_TOKEN -> "YAML_STREAM_END_TOKEN"
  | YAML_VERSION_DIRECTIVE_TOKEN (_,_) -> "YAML_VERSION_DIRECTIVE_TOKEN (_,_)"
  | YAML_TAG_DIRECTIVE_TOKEN (_,_) -> "YAML_TAG_DIRECTIVE_TOKEN (_,_)"
  | YAML_DOCUMENT_START_TOKEN -> "YAML_DOCUMENT_START_TOKEN"
  | YAML_DOCUMENT_END_TOKEN -> "YAML_DOCUMENT_END_TOKEN"
  | YAML_BLOCK_SEQUENCE_START_TOKEN -> "YAML_BLOCK_SEQUENCE_START_TOKEN"
  | YAML_BLOCK_MAPPING_START_TOKEN -> "YAML_BLOCK_MAPPING_START_TOKEN"
  | YAML_BLOCK_END_TOKEN -> "YAML_BLOCK_END_TOKEN"
  | YAML_FLOW_SEQUENCE_START_TOKEN -> "YAML_FLOW_SEQUENCE_START_TOKEN"
  | YAML_FLOW_SEQUENCE_END_TOKEN -> "YAML_FLOW_SEQUENCE_END_TOKEN"
  | YAML_FLOW_MAPPING_START_TOKEN -> "YAML_FLOW_MAPPING_START_TOKEN"
  | YAML_FLOW_MAPPING_END_TOKEN -> "YAML_FLOW_MAPPING_END_TOKEN"
  | YAML_BLOCK_ENTRY_TOKEN -> "YAML_BLOCK_ENTRY_TOKEN"
  | YAML_FLOW_ENTRY_TOKEN -> "YAML_FLOW_ENTRY_TOKEN"
  | YAML_KEY_TOKEN -> "YAML_KEY_TOKEN"
  | YAML_VALUE_TOKEN -> "YAML_VALUE_TOKEN"
  | YAML_ALIAS_TOKEN _ -> "YAML_ALIAS_TOKEN _"
  | YAML_ANCHOR_TOKEN _ -> "YAML_ANCHOR_TOKEN _"
  | YAML_TAG_TOKEN (_,_) -> "YAML_TAG_TOKEN (_,_)"
  | YAML_SCALAR_TOKEN (_,_) -> "YAML_SCALAR_TOKEN (_,_)"

type yaml_parser
external open_parser  : string -> yaml_parser = "open_parser"
external close_parser : yaml_parser -> unit = "close_parser"
external next_token   : yaml_parser -> token = "next_token"
external get_error    : yaml_parser -> (yaml_error_type * string * string) = "get_error"

let rec parse_sequence p =
    let rec scan () =
        match next_token p  with
        | YAML_BLOCK_END_TOKEN -> []
        | YAML_BLOCK_ENTRY_TOKEN ->
            let entry = parse_yaml (next_token p) p in
            entry :: scan ()
        | token  -> failwith ("Unexpected YAML token in sequence: " ^ string_of_token token)
        in
    Collection ( scan () |> List.rev )

and parse_sequence_flow p =
    (* In a flow sequence, we expect one entry before the first entry token. *)
    match next_token p with
    | YAML_FLOW_SEQUENCE_END_TOKEN -> Collection []
    | token ->
      let first_entry = parse_yaml token p in
      let rec scan () =
          match next_token p  with
          | YAML_FLOW_SEQUENCE_END_TOKEN -> []
          | YAML_FLOW_ENTRY_TOKEN ->
              let entry = parse_yaml (next_token p) p in
              entry :: scan ()
          | token  -> failwith ("Unexpected YAML token in flow sequence: " ^ string_of_token token)
          in
      Collection ( first_entry :: (scan () |> List.rev) )

and parse_mapping p =
    let rec scan key () = match next_token p, key with
        | YAML_BLOCK_END_TOKEN, None -> []
        | YAML_KEY_TOKEN,       None ->
            let key   = parse_yaml (next_token p) p in
            scan (Some key) ()
        | YAML_VALUE_TOKEN,     Some key ->
            let value = parse_yaml (next_token p) p in
            ( key, value)  :: scan None ()
        | YAML_BLOCK_END_TOKEN, Some _ -> failwith "Unmatched key token."
        | YAML_KEY_TOKEN,       Some _ -> failwith "Two key tokens in a row."
        | YAML_VALUE_TOKEN,     None   -> failwith "Value token without a key"
        | token , _ -> failwith ("Unexpected YAML token in mapping: " ^ string_of_token token)
        in
    Structure ( scan None () |> List.rev )

and parse_yaml token p =
    match token  with
    | YAML_STREAM_START_TOKEN _ -> parse_yaml (next_token p) p
    | YAML_BLOCK_SEQUENCE_START_TOKEN -> parse_sequence p
    | YAML_FLOW_SEQUENCE_START_TOKEN -> parse_sequence_flow p
    | YAML_BLOCK_MAPPING_START_TOKEN  -> parse_mapping p
    | YAML_SCALAR_TOKEN ( _ , v ) -> Scalar v
    | YAML_NO_TOKEN ->  begin
        match get_error p with
            | YAML_NO_ERROR, _ , _ -> failwith "Unexpected end of YAML stream"
            | _ , errmsg, ctx -> failwith ("YAML parsing error: " ^ errmsg ^ ". Context: " ^ ctx)
        end
    | _ -> failwith ("Unexpected YAML token: " ^ string_of_token token)

let load filename =
    let p = open_parser filename in
    let yaml = parse_yaml (next_token p) p in
    let () = close_parser p in
    yaml
